ssz/src/test/java/org/ethereum/beacon/ssz/SSZTypeTest.java

```diff
@@ -0,0 +1,539 @@
+  @Test
+  public void testTypeResolver2() throws Exception {
+    SSZBuilder sszBuilder = new SSZBuilder()
+        .withExternalVarResolver(s -> "testSize".equals(s) ? 1 : null);
+    SSZSerializer serializer = sszBuilder.buildSerializer();
+
+    SSZType sszType = sszBuilder.getTypeResolver().resolveSSZType(Impl2.class);
+    System.out.println(dumpType(sszType, ""));
+
+    Assert.assertTrue(sszType instanceof SSZContainerType);
+    SSZContainerType containerType = (SSZContainerType) sszType;
+
+    Assert.assertEquals(2, containerType.getChildTypes().size());
+    Assert.assertEquals("a1", containerType.getChildTypes().get(0).getTypeDescriptor().getName());
+    Assert.assertEquals("a2", containerType.getChildTypes().get(1).getTypeDescriptor().getName());
+
+    byte[] bytes1 = serializer.encode(new Impl2());
+    System.out.println(BytesValue.wrap(bytes1));
+
+    byte[] bytes2 = serializer.encode(new Impl1());
+    System.out.println(BytesValue.wrap(bytes2));
+    Assert.assertArrayEquals(bytes1, bytes2);
+    Assert.assertTrue(BytesValue.wrap(bytes1).toString().contains("1111"));
+    Assert.assertTrue(BytesValue.wrap(bytes1).toString().contains("2222"));
+  }
+
+  @SSZSerializable
+  public static class H1 {
+    @SSZ public int a1;
+    @SSZ public long a2;
+    @SSZ public int a3;
+  }
+
+  @SSZSerializable
+  public static class H2 {
+    @SSZ public int a1;
+    @SSZ public long a2;
+  }
+
+  @Test
+  public void testHashTruncated1() throws Exception {
+    SSZHasher hasher = new SSZBuilder().buildHasher(Hashes::keccak256);
+
+    H1 h1 = new H1();
+    h1.a1 = 0x1111;
+    h1.a2 = 0x2222;
+    h1.a3 = 0x3333;
+
+    H2 h2 = new H2();
+    h2.a1 = 0x1111;
+    h2.a2 = 0x2222;
+
+    byte[] h1h = hasher.hash(h1);
+    byte[] h2h = hasher.hash(h2);
+    byte[] h1hTrunc = hasher.hashTruncateLast(h1, H1.class);
+
+    Assert.assertArrayEquals(h2h, h1hTrunc);
+  }
+
+  @SSZSerializable
+  public static class I1 implements ObservableComposite {
+    UpdateListener updateListener;
+
+    @SSZ private int a1;
+    @SSZ private long a2;
+    @SSZ private int a3;
+
+    public I1(int a1, long a2, int a3) {
+      this.a1 = a1;
+      this.a2 = a2;
+      this.a3 = a3;
+    }
+
+    @Override
+    public UpdateListener getUpdateListener(
+        String observerId, Supplier<UpdateListener> listenerFactory) {
+
+      return updateListener != null ? updateListener : (updateListener = listenerFactory.get());
+    }
+
+    public int getA1() {
+      return a1;
+    }
+
+    public long getA2() {
+      return a2;
+    }
+
+    public int getA3() {
+      return a3;
+    }
+
+    public void setA1(int a1) {
+      this.a1 = a1;
+      updateListener.childUpdated(0);
+    }
+
+    public void setA2(long a2) {
+      this.a2 = a2;
+      updateListener.childUpdated(1);
+    }
+
+    public void setA3(int a3) {
+      this.a3 = a3;
+      updateListener.childUpdated(2);
+    }
+  }
+
+  @Test
+  public void testHashIncremental1() throws Exception {
```

Comment:
I'd move it to separate `SSZIncrementalHashTest` to make it clear that there is a test verifying matching of simple hash and incremental